File: VB/Src/frmIAVB.vb

' Fichier frmIAVB.vb
' ------------------

Imports System.Collections.Generic
Imports System.Speech
Imports System.Speech.Synthesis

Namespace IAVB

Friend Class frmIAVB : Inherits Form

#Region "Configuration"

' Booléen pour forcer le fonctionnement dans la version originale (1984)
'  ou bien modifiée (2001)
Private Const bVersionModifiee As Boolean = True ' True par défaut

' Normaliser les espaces et sauts de ligne en sortie
Private Const bNormalisationSortieTrimEtVbLf As Boolean = True ' True par défaut

' Convertir tout en minuscules (pour éviter d'avoir à taper des majuscules avec accent,
'  et pour pouvoir prononcer "ça" via MS-Agent)
Private Const bTraiterEnMinuscules As Boolean = True ' True par défaut

' 29/04/2017 Enlever les accents (pour pouvoir comparer avec la version originale de 1984)
Private Const bTraiterSansAccents As Boolean = False ' False par défaut

#End Region

#Region "Initialisations"

Private Const iCodeToucheEntree% = 13
Private Const iNumLigneDebutExemples% = 8 ' Pour la démo sur les exemples

Private m_bMSAgent As Boolean

' Pour pouvoir recopier la ligne ListIA dans la fonction RecopierLigneListIA
Private m_bPositionnement As Boolean

Private m_oIAVB As New IAVB.clsIAVB

' SpeechSynthesizer Class Provides access to the functionality of an installed a speech synthesis engine.
Private WithEvents m_synth As New SpeechSynthesizer
Const sVoixFR$ = "Microsoft Hortense Desktop"
Const sVoix_enUSWin10$ = "Microsoft Zira Desktop"
Const sVoix_enUSWin7$ = "Microsoft Anna"
Const sFrancais$ = "Français (Hortense)"
Const sAnglais$ = "Anglais (Zira)"
Private m_sVoix$ = ""

Const sModeVocalDef$ = sFrancais
Const sModeSilencieux$ = "Silencieux"
Const sModeMSAgent$ = "MS-Agent"
Const iIndexModeSilencieux% = 0
'Const iIndexModeMSAgent% = 1
Dim iIndexModeSynthVocWin% = 2

Const sTxtCmdSuiv$ = "->"
Const sMsgCmdSuiv$ = "Exemple suivant"
Const sTxtCmdSuivStop$ = "St."
Const sMsgCmdSuivStop$ = "Stop"

Private Sub Initialisations()

    Dim sTxt$ = Me.Text & " - V" & sVersionAppli & " (" & sDateVersionAppli & ")"
    If bDebug Then sTxt &= " - Debug"
    Dim b64 As Boolean = Is64BitProcess()
    If b64 Then sTxt &= " - 64 bits" Else sTxt &= " - 32 bits"
    Me.Text = sTxt

    Dim bVoixActive As Boolean
    If Not b64 Then VerifierInstallAgent(m_bMSAgent, bVoixActive)

    iIndexModeSynthVocWin = 1
    Me.listParole.Items.Add(sModeSilencieux)
    If m_bMSAgent Then Me.listParole.Items.Add(sModeMSAgent) : iIndexModeSynthVocWin = 2
    InitVoix()
    Me.listParole.SelectedIndex = iIndexModeSilencieux

    m_oIAVB.Initialiser(bVersionModifiee, bNormalisationSortieTrimEtVbLf, _
        bTraiterEnMinuscules, bTraiterSansAccents)
    InitAssertionsExemples()

End Sub

Private Sub InitVoix()

    ' SpeechSynthesizer.GetInstalledVoices, méthode (CultureInfo)
    ' https://msdn.microsoft.com/fr-fr/library/ms586870(v=vs.110).aspx?cs-save-lang=1&cs-lang=vb#code-snippet-1
    ' Output information about all of the installed voices that
    '  support the en-US locacale. 
    Dim dicoVoix As New Dictionary(Of String, VoiceInfo)
    For Each voix As InstalledVoice In m_synth.GetInstalledVoices() 'New CultureInfo("en-US"))
        Dim info As VoiceInfo = voix.VoiceInfo
        Dim sCle$ = info.Name & " (" & info.Culture.Name & ", " & info.Gender.ToString() & ")"
        ' Remplacer les noms des deux voix standards de Windows 10 par Français et Anglais
        If info.Name = sVoixFR Then sCle = sFrancais
        If info.Name = sVoix_enUSWin10 Then sCle = sAnglais
        If info.Name = sVoix_enUSWin7 Then sCle = sAnglais
        If Not dicoVoix.ContainsKey(sCle) Then
            dicoVoix.Add(sCle, info)
            Me.listParole.Items.Add(sCle)
        End If
    Next

End Sub

Private Sub listParole_SelectedIndexChanged(sender As Object, e As EventArgs) _
    Handles listParole.SelectedIndexChanged

    If listParole.Text = sModeMSAgent Then
        If Not bInitMSAgent(m_bMSAgent, m_oIAVB.m_bVoixActive) Then
            listParole.Text = sModeSilencieux
        End If
    Else
        bMasquerMSAgent()
    End If

    If listParole.Text = sModeSilencieux Then
        m_oIAVB.m_bVoixActive = False
    Else
        m_oIAVB.m_bVoixActive = True
    End If

    If listParole.SelectedIndex >= iIndexModeSynthVocWin Then
        Dim sLangue$ = listParole.Text
        Dim sVoix$ = ""
        If sLangue = sFrancais Then
            sVoix = sVoixFR
        ElseIf sLangue = sAnglais Then
            sVoix = sVoix_enUSWin10
        Else
            sVoix = sLangue
            Dim iPos% = sVoix.IndexOf("(")
            If iPos > -1 Then sVoix = sVoix.Substring(0, iPos - 1).Trim
        End If
        m_sVoix = sVoix
    End If

End Sub

Public Function bDire(sParole$) As Boolean

    If Me.listParole.Text = sModeMSAgent Then
        ' Les minuscules améliorent la prononciation des MS-Agent
        If m_bMSAgent AndAlso bTraiterEnMinuscules Then sParole = sParole.ToLower
        Return bDireMSAgent(sParole)
    End If

    If listParole.Text = sModeSilencieux Then Return False

    If Not IsNothing(m_synth) AndAlso m_sVoix.Length > 0 Then

        While m_synth.State = SynthesizerState.Speaking
            Attendre()
        End While

        If m_synth.State = SynthesizerState.Ready Then
            m_synth.SelectVoice(m_sVoix)
            Activation(bDesactiver:=True)
            'm_synth.SpeakAsync(sParole)
            m_synth.Speak(sParole) ' Attendre la fin
            Activation()
        End If
    Else
        listParole.Text = sModeSilencieux
    End If

    Return True

End Function

' Si SpeakAsync
'Private Sub MyEventHandler( eventSender As Object, _
'    eventArgs As SpeakCompletedEventArgs) Handles m_synth.SpeakCompleted
'    Activation()
'End Sub

Private Sub Activation(Optional bDesactiver As Boolean = False)

    ' Non, car on ne peut plus réactiver Me.chkAuto ou faire Stop
    'Me.Enabled = bDesactiver

    Me.chkAuto.Enabled = True ' Tjrs actif
    Me.cmdSuivant.Enabled = True ' Tjrs actif (Stop le mode Auto)

    Me.listAssert.Enabled = Not bDesactiver
    Me.listIA.Enabled = Not bDesactiver
    Me.textInput.Enabled = Not bDesactiver
    Me.cmdGo.Enabled = Not bDesactiver
    Me.cmdInstall.Enabled = Not bDesactiver

End Sub

Private Sub AjouterExemple(sExemple$)
    Me.listAssert.Items.Add(sExemple)
End Sub

Private Sub InitAssertionsExemples()

    AjouterExemple(m_oIAVB.sCmdEffBase & " ' Effacer toute la base")
    AjouterExemple(m_oIAVB.sCmdEff & " ' Effacer la denière assertion")
    AjouterExemple(m_oIAVB.sCmdEff & "1 ' Effacer l'assertion n°1")
    AjouterExemple(m_oIAVB.sCmdLister & " ' Lister les assertions de la base")
    AjouterExemple(m_oIAVB.sCmdCopier & " ' Copier la discussion dans le presse papier")
    AjouterExemple(m_oIAVB.sCmdSilence & " ' Mettre en sourdine (autre poss.: masquer l'agent)")
    AjouterExemple(m_oIAVB.sCmdParler & " ' Ré-activer la voix")

    Dim lst As New List(Of String)
    Dim ex As New IAVB.clsInitIAVB(lst)
    ex.InitAssertionsExemples()
    For Each sExemple As String In lst
        AjouterExemple(sExemple)
    Next

End Sub

#End Region

Private Sub TraiterCmd()

    Dim bMemVoixActive As Boolean = m_oIAVB.m_bVoixActive

    ' Eviter la réentrance dans les fonctions
    If m_oIAVB.m_bVoixActive Then Activation(bDesactiver:=True)

    ' Si la zone de saisie est multiligne, faire un Trim
    Dim sTxt$ = textInput.Text.Trim()
    m_oIAVB.IAVBMain(sTxt)

    If m_oIAVB.m_bCopierPressePapier Then
        If bCopierPressePapier(m_oIAVB.m_sDiscussion) Then
            m_oIAVB.CopiePressePapierOk()
        Else
            m_oIAVB.CopiePressePapierEchec()
        End If
    End If

    If Me.textInput.Text.Length = 0 Then
        Me.listIA.Items.Add("")
        PositionnerListIA()
    End If

    ' Activer ou réactiver la parole
    If Not bMemVoixActive AndAlso m_oIAVB.m_bVoixActive Then listParole.Text = sModeVocalDef

    Dim sReponseVocale$ = m_oIAVB.m_sReponseVocale
    If sReponseVocale.Length > 0 Then

        Dim asReponses$() = sReponseVocale.Split(vbLf.ToCharArray)
        Dim i% = 0
        For Each sReponse As String In asReponses
            If sReponse.Trim.Length = 0 Then Continue For
            ' 12/05/2017
            If i = 0 AndAlso asReponses.Length >= 1 AndAlso _
               asReponses(1) = asReponses(0) Then
                ' Rappel de la question : dans ce cas on peut afficher le texte en 1er

                ' Réponse texte
                Me.listIA.Items.Add(sReponse)
                PositionnerListIA()

                bDire(sReponse) ' Réponse vocale

            ElseIf i = 1 AndAlso asReponses.Length >= 1 AndAlso _
               asReponses(1) = asReponses(0) Then
                ' Déjà traité
            Else
                If i Mod 2 = 0 Then
                    bDire(sReponse) ' Réponse vocale
                Else
                    ' Réponse texte
                    Me.listIA.Items.Add(sReponse)
                    PositionnerListIA()
                End If
            End If
            i += 1
        Next

        ' Désactiver la parole
        If Not m_oIAVB.m_bVoixActive Then listParole.Text = sModeSilencieux

        GoTo Fin

    End If

    If m_oIAVB.m_sRappelQuestion.Length > 0 Then
        Me.listIA.Items.Add(m_oIAVB.m_sRappelQuestion)
        PositionnerListIA()
    End If

    If m_oIAVB.m_sReponse.Length > 0 Then
        Dim asReponses$() = m_oIAVB.m_sReponse.Split(vbLf.ToCharArray)
        For Each sReponse As String In asReponses
            If sReponse.Trim.Length = 0 Then Continue For
            Me.listIA.Items.Add(sReponse)
        Next
        PositionnerListIA()
    End If

Fin:
    If bMemVoixActive Then
        Activation()
        cmdSuivant.Focus()
    End If

End Sub

Private Sub RecopierLigneListIA()
    If Not m_bPositionnement Then textInput.Text = listIA.Text
End Sub

Public Sub PositionnerListIA()
    ' Toujours se positionner sur la dernière ligne
    m_bPositionnement = True
    listIA.SelectedIndex = listIA.Items.Count - 1
    m_bPositionnement = False
End Sub

Private Sub TraiterAssertion()

    textInput.Text = listAssert.Text

    ' Veiller à ce que la question soit bien affichée avant d'activer la synthèse vocale
    TraiterMsgSysteme_DoEvents()

    TraiterCmd()

End Sub

Private Sub AssertionSuivante()
    If listAssert.SelectedIndex < iNumLigneDebutExemples Then
        listAssert.SelectedIndex = iNumLigneDebutExemples
    ElseIf listAssert.SelectedIndex < listAssert.Items.Count - 1 Then
        listAssert.SelectedIndex = listAssert.SelectedIndex + 1
    End If
    TraiterAssertion()
End Sub

#Region "Gestion des événements"

Private Sub frmIAVB_Load(eventSender As Object, eventArgs As EventArgs) Handles MyBase.Load
    Initialisations()
End Sub

Private Sub cmdGo_Click(eventSender As Object, eventArgs As EventArgs) Handles cmdGo.Click
    TraiterCmd()
End Sub

Private Sub chkAuto_Click(sender As Object, e As EventArgs) Handles chkAuto.Click
    If Me.chkAuto.Checked Then
        'Me.cmdSuivant.Text = sTxtCmdSuivStop
        'Me.ToolTip1.SetToolTip(Me.cmdSuivant, sMsgCmdSuivStop)
    Else
        Me.cmdSuivant.Text = sTxtCmdSuiv
        Me.ToolTip1.SetToolTip(Me.cmdSuivant, sMsgCmdSuiv)
    End If
End Sub

Private Sub cmdSuivant_Click(eventSender As Object, eventArgs As EventArgs) _
    Handles cmdSuivant.Click

    Dim sMemAssertion2$, sMemAssertion$
    sMemAssertion = ""
    Static bEnCours As Boolean = False
    If bEnCours Then GoTo Fin

Recommencer:
    bEnCours = True
    sMemAssertion2 = sMemAssertion
    sMemAssertion = Me.listIA.Text.Trim
    AssertionSuivante()

    If Me.chkAuto.Checked Then

        Me.cmdSuivant.Text = sTxtCmdSuivStop
        Me.ToolTip1.SetToolTip(Me.cmdSuivant, sMsgCmdSuivStop)

        If Me.listIA.Text.Trim.Length = 0 AndAlso _
           sMemAssertion.Length = 0 AndAlso sMemAssertion2.Length = 0 Then
            Me.textInput.Text = m_oIAVB.sCmdCopier ' "/C"
            TraiterCmd()
            bEnCours = False
            Exit Sub
        End If
        ' Voir si l'utilisateur à cliquer sur Stop
        TraiterMsgSysteme_DoEvents()
        GoTo Recommencer
    End If

Fin:
    bEnCours = False
    Me.chkAuto.Checked = False
    Me.cmdSuivant.Text = sTxtCmdSuiv
    Me.ToolTip1.SetToolTip(Me.cmdSuivant, sMsgCmdSuiv)

End Sub

Private Sub ListAssert_SelectedIndexChanged(eventSender As Object, _
    eventArgs As EventArgs) Handles listAssert.SelectedIndexChanged
    textInput.Text = listAssert.Text
End Sub

Private Sub ListAssert_DoubleClick(eventSender As Object, eventArgs As EventArgs) _
    Handles listAssert.DoubleClick
    TraiterAssertion()
End Sub

Private Sub listIA_SelectedIndexChanged(eventSender As Object, eventArgs As EventArgs) _
    Handles listIA.SelectedIndexChanged
    RecopierLigneListIA()
End Sub

Private Sub listIA_DoubleClick(eventSender As Object, eventArgs As EventArgs) _
    Handles listIA.DoubleClick
    RecopierLigneListIA()
    TraiterCmd()
End Sub

Private Sub textInput_KeyPress(eventSender As Object, eventArgs As KeyPressEventArgs) _
    Handles textInput.KeyPress

    Dim iCodeTouche% = Asc(eventArgs.KeyChar)
    If iCodeTouche = iCodeToucheEntree Then ' Touche Entrée
        TraiterCmd()
        Exit Sub
    End If

End Sub

Private Sub cmdInstall_Click(sender As Object, e As EventArgs) Handles cmdInstall.Click
    bInstallationMSAgent(m_bMSAgent, m_oIAVB.m_bVoixActive)
End Sub

#End Region

End Class

End Namespace

File: CSharp/WinForm/frmIAVB.cs

using IAVB; // pour clsIAVB2.vbLf
//using LibIAVB2.clsIAVB2; // pour vbLf : Ne marche pas

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

using System.Speech;
using System.Speech.Synthesis;

//using System.Diagnostics; // Pour Debug.Write()

namespace IAVB
{
    public partial class frmIAVB : Form
    {
        private bool bDebug = IAVB.Program.bDebug;

        // Pour la démo sur les exemples
        private const int iNumLigneDebutExemples = 8;

        private const bool bVersionModifiee = true;
        private const bool bNormalisationSortieTrimEtVbLf = true;

        public bool m_bVoixActive;

        private SpeechSynthesizer m_synth = new SpeechSynthesizer();

        private IAVB.clsIAVB m_oIAVB = new IAVB.clsIAVB();

        #region "Initialisations"

        public frmIAVB()
        {
            InitializeComponent();
            if (bDebug) this.StartPosition = FormStartPosition.CenterScreen;
        }

        private void frmIAVB_Load(object sender, EventArgs e)
        {
            Initialisations();
        }

        private void Initialisations()
        {
            string sTxt = Text + " - V" + IAVB.Program.sVersionAppli + 
                " (" + IAVB.Program.sDateVersionAppli + ")";
            if (bDebug) sTxt += " - Debug";
            this.Text = sTxt;

            Utils.clsAppUtils.AppTitle = clsIAVB.sIAVB;

            listParole.Items.Add(clsSynthVoc.sModeSilencieux);
            InitVoix();
            listParole.SelectedIndex = clsSynthVoc.iIndexModeSilencieux;

            m_oIAVB.Initialiser(bVersionModifiee, bNormalisationSortieTrimEtVbLf);
            InitAssertionsExemples();
        }
        
        private void InitVoix()
        {
            var lstParoles = new List<string>();
            clsSynthVoc.InitVoix(m_synth, lstParoles);
            foreach(string sParole in lstParoles)
            {
                listParole.Items.Add(sParole);
            }
        }

        private void AjouterExemple(string sExemple)
        {
            listAssert.Items.Add(sExemple);
        }
                    
        private void InitAssertionsExemples()
        {
            AjouterExemple(m_oIAVB.sCmdEffBase + " ' Effacer toute la base");
            AjouterExemple(m_oIAVB.sCmdEff + " ' Effacer la denière assertion");
            AjouterExemple(m_oIAVB.sCmdEff + "1 ' Effacer l'assertion n°1");
            AjouterExemple(m_oIAVB.sCmdLister + " ' Lister les assertions de la base");
            AjouterExemple(m_oIAVB.sCmdCopier + " ' Copier la discussion dans le presse papier");
            AjouterExemple(m_oIAVB.sCmdSilence + " ' Mettre en sourdine");
            AjouterExemple(m_oIAVB.sCmdParler + " ' Ré-activer la voix");

            List<string> lst = new List<string>();
            IAVB.clsInitIAVB ex = new IAVB.clsInitIAVB(lst);
            ex.InitAssertionsExemples();
            foreach (string sExemple in lst)
            {
                AjouterExemple(sExemple);
            }

        }

        #endregion
        
        private void TraiterAssertion()
        {
            textInput.Text = listAssert.Text;
            
            // Veiller à ce que la question soit bien affichée avant d'activer la synthèse vocale
            Application.DoEvents();

            TraiterCmd();
        }

        private void TraiterCmd()
        {
            bool bMemVoixActive = m_oIAVB.m_bVoixActive;

            // Eviter la réentrance dans les fonctions
            if (m_oIAVB.m_bVoixActive) Activation(bDesactiver: true);

            // Si la zone de saisie est multiligne, faire un Trim
            string sTxt = textInput.Text.Trim();
            m_oIAVB.IAVBMain(sTxt);

            if (m_oIAVB.m_bCopierPressePapier) 
            {
                if (Utils.clsMessageUtils.bCopyToClipboard(m_oIAVB.m_sDiscussion))
                    m_oIAVB.CopierPressePapierOk();
                else
                    m_oIAVB.CopierPressePapierPb();
            }

            if (textInput.Text.Length == 0)
            {
                listIA.Items.Add("");
                PositionnerListIA();
            }

            // Activer ou réactiver la parole
            if (!bMemVoixActive && m_oIAVB.m_bVoixActive) listParole.Text = clsSynthVoc.sModeVocalDef;

            string sReponseVocale = m_oIAVB.m_sReponseVocale;
            if (sReponseVocale.Length > 0)
            {
                string[] asReponses = sReponseVocale.Split(
                    new string[] { clsIAVB.vbLf }, StringSplitOptions.None);
                int i = 0;
                foreach (string sReponse in asReponses)
                {
                    if (sReponse.Trim().Length == 0) continue;
                    if (i == 0 && asReponses.Length >= 1 && asReponses[1] == asReponses[0]) 
                    {
                        // Rappel de la question : dans ce cas on peut afficher le texte en 1er

                        // Réponse texte
                        if (sReponse.Trim().Length == 0) continue;
                        listIA.Items.Add(sReponse);
                        PositionnerListIA();
                        
                        bDire(sReponse); // Réponse vocale
                    }
                    else if (i == 1 && asReponses.Length >= 1 && asReponses[1] == asReponses[0])
                    {
                        // Déjà traité
                    }
                    else if (i % 2 == 0) bDire(sReponse); // Réponse vocale
                    else
                    {
                        // Réponse texte
                        if (sReponse.Trim().Length == 0) continue;
                        listIA.Items.Add(sReponse);
                        PositionnerListIA();
                    }
                    i += 1;
                }

                // Désactiver la parole
                if (!m_oIAVB.m_bVoixActive) listParole.Text = clsSynthVoc.sModeSilencieux;

                goto Fin;
            }

            if (m_oIAVB.m_sRappelQuestion.Length > 0)
            {
                listIA.Items.Add(m_oIAVB.m_sRappelQuestion);
                PositionnerListIA();
            }

            if (m_oIAVB.m_sReponse.Length > 0)
            {
                string[] asReponses = m_oIAVB.m_sReponse.Split(
                    new string[] { clsIAVB.vbLf }, StringSplitOptions.None);
                foreach (string sReponse in asReponses)
                {
                    if (sReponse.Trim().Length == 0)
                        continue;
                    listIA.Items.Add(sReponse);
                }
                PositionnerListIA();
            }

        Fin:
            if (bMemVoixActive)
            {
                Activation();
                cmdSuivant.Focus();
            }

        }
        
        public bool bDire(string sParole)
        {
            //if (listParole.Text == sModeMSAgent) return bDireMSAgent(sParole);

            if (listParole.Text == clsSynthVoc.sModeSilencieux) return false;

            if ((!(m_synth == null) && (clsSynthVoc.m_sVoix.Length > 0))) 
            {
                while (m_synth.State == SynthesizerState.Speaking) Util.Attendre();
            
                if (m_synth.State == SynthesizerState.Ready) {
                    m_synth.SelectVoice(clsSynthVoc.m_sVoix);
                    Activation(bDesactiver:true);
                    // m_synth.SpeakAsync(sParole)
                    m_synth.Speak(sParole);
                    //  Attendre la fin
                    Activation();
                }
            
            }
            else listParole.Text = clsSynthVoc.sModeSilencieux;
        
            return true;
        }
        
        private void Activation(bool bDesactiver= false) 
        {
            //chkAuto.Enabled = true;
            cmdSuivant.Enabled = true;
            listAssert.Enabled = !bDesactiver;
            listIA.Enabled = !bDesactiver;
            textInput.Enabled = !bDesactiver;
            cmdGo.Enabled = !bDesactiver;

            // Ici DoEvents permet d'afficher la répétition de la question avant la réponse d'IAVB
            Application.DoEvents(); // Requis en C#, pas en VB
        }

        public void PositionnerListIA()
        {
            // Toujours se positionner sur la dernière ligne
            listIA.SelectedIndex = listIA.Items.Count - 1;
        }

        #region "Gestion des événements"

        private void cmdGo_Click(object sender, EventArgs e)
        {
            TraiterCmd();
        }

        private void cmdSuivant_Click(object eventSender, EventArgs eventArgs)
        {
            if (listAssert.SelectedIndex < iNumLigneDebutExemples)
                listAssert.SelectedIndex = iNumLigneDebutExemples;
            else if (listAssert.SelectedIndex < listAssert.Items.Count - 1)
                listAssert.SelectedIndex = listAssert.SelectedIndex + 1;
            TraiterAssertion();
        }

        private void listAssert_SelectedIndexChanged(object sender, EventArgs e)
        {
            textInput.Text = listAssert.Text;
        }

        private void listAssert_DoubleClick(object sender, EventArgs e)
        {
            TraiterAssertion();
        }
        
        private void listIA_SelectedIndexChanged(object sender, EventArgs e)
        {
            textInput.Text = listIA.Text;
        }

        private void listIA_DoubleClick(object sender, EventArgs e)
        {
            textInput.Text = listIA.Text;
            TraiterCmd();
        }

        void textInput_KeyPress(object sender, System.Windows.Forms.KeyPressEventArgs e)
        {
            if (e.KeyChar == Convert.ToChar(Keys.Enter)) TraiterCmd();
        }

        private void listParole_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (listParole.Text == clsSynthVoc.sModeSilencieux) m_oIAVB.m_bVoixActive = false;
            else m_oIAVB.m_bVoixActive = true;
            clsSynthVoc.LireVoix(listParole.SelectedIndex, listParole.Text);
        }

        #endregion

    }
}
